### PR TITLE
Enhance: add http server callbacks

### PR DESCRIPTION
**Pull request by xqbumu**
in order to set or reset some build config, i have added some callbacks, so we can reuse HTTPServerBuilder without copy.

thanks for your greats jobs.

### CODE REVIEW

File: constructors/service.go [hunk 1]
```diff
@@ -56,10 +57,16 @@ func (deps webServiceDependencies) pingService(ctx context.Context, service serv
 	}
 	return
 }
+
 func (deps webServiceDependencies) getGRPCAddress(ports []server.ListenInfo) string {
 	for _, info := range ports {
 		if info.Type == server.GRPCServer {
-			return info.Address
+			switch info.Network {
```
Comment (talgendler) [inline constructors/service.go:64]: Why do you need to add a schema to Address ? It's not an URI

File: interfaces/http/server/interfaces.go [hunk 2]
```diff
@@ -23,6 +23,7 @@ const (
 
 // ListenInfo defines port info
 type ListenInfo struct {
+	Network string        `json:"network"`
```
Comment (talgendler) [inline interfaces/http/server/interfaces.go:26]: What is the use case here?

**Review by talgendler — COMMENTED** [review]
Hi @xqbumu thank you for your contribution. 
Can you please explain why do you need to mutate the builder ? Builders and especially that one are called once per Application during setup. 
Regardless there were no tests added.
In addition you added other changes related to network addresses, why ?

File: interfaces/http/server/interfaces.go (hunk 2 above)
Comment (xqbumu) [reply interfaces/http/server/interfaces.go:26]: The 'Network' field is used by getGRPCAddress to decide which address format should be return.

File: constructors/service.go (hunk 1 above)
Comment (xqbumu) [reply constructors/service.go:64]: It used by grpc dialer when we change the grpc server with unix socket in the callbacks in the PR.

we can get more detail in https://github.com/grpc/grpc/blob/master/doc/naming.md

**Comment by xqbumu** [comment]
> Hi @xqbumu thank you for your contribution. Can you please explain why do you need to mutate the builder ? Builders and especially that one are called once per Application during setup. Regardless there were no tests added. In addition you added other changes related to network addresses, why ?

we have used the 'mortar' in out project, but in some reasons, the export port is offered by another lib which the port maybe dynamic with another protocol, so we use SetCustomServer to export external handler.

you can find our code here: https://github.com/airdb/xadmin-api/blob/dev/app/mortar/serverless.go#L122

at the same time, we also want reduce the port in order to cut down the ports used by replace with unix socket.

**Comment by talgendler** [comment]
I see, can I suggest another option ?
You can change the ports in the [config file](https://github.com/airdb/xadmin-api/blob/main/config/config.yml#L11)
```yaml
mortar:
 server:
  host: localhost
  grpc:
   port: 5380
  rest:
   external:
    port: 5381
   internal:
    port: 5382
```

OR

If you want this to be dynamic, change some code here: https://github.com/airdb/xadmin-api/blob/main/app/mortar/config.go#L15
Instead of simply returning the Config interface, make some changes and then return. Since this is a constructor everything depend on it and will have your new values.
```go
func ViperFxOption(configFilePath string, additionalFilePaths ...string) fx.Option {
	return fx.Provide(func() (cfg.Config, error) {
		builder := bviper.Builder().SetConfigFile(configFilePath)
		for _, extraFile := range additionalFilePaths {
			builder = builder.AddExtraConfigFile(extraFile)
		}
		config, err := builder.Build()
		config.Set("mortar.server.grpc.port", 2345)
		return config, err
	})
}
```

**Comment by xqbumu** [comment]
> I see, can I suggest another option ? You can change the ports in the [config file](https://github.com/airdb/xadmin-api/blob/main/config/config.yml#L11)
> 
> ```yaml
> mortar:
>  server:
>   host: localhost
>   grpc:
>    port: 5380
>   rest:
>    external:
>     port: 5381
>    internal:
>     port: 5382
> ```
> 
> OR
> 
> Override using ENVIRONMENT Variable `MORTAR_SERVER_GRPC_PORT`
> 
> OR
> 
> If you want this to be dynamic, change some code here: https://github.com/airdb/xadmin-api/blob/main/app/mortar/config.go#L15 Instead of simply returning the Config interface, make some changes and then return. Since this is a constructor everything depend on it and will have your new values.
> 
> ```go
> func ViperFxOption(configFilePath string, additionalFilePaths ...string) fx.Option {
> 	return fx.Provide(func() (cfg.Config, error) {
> 		builder := bviper.Builder().SetConfigFile(configFilePath)
> 		for _, extraFile := range additionalFilePaths {
> 			builder = builder.AddExtraConfigFile(extraFile)
> 		}
> 		config, err := builder.Build()
> 		config.Set("mortar.server.grpc.port", 2345)
> 		return config, err
> 	})
> }
> ```

thanks for your options. they are valid methods to solved dynamic port, but there is a protocol question here

https://github.com/airdb/xadmin-api/blob/dev/app/mortar/serverless.go#L59
https://github.com/serverless-plus/tencent-serverless-go/blob/master/faas/entry.go#L62

and it wraps http handler under the rpc call,

so how to use dynamic port in fx is not the main question, the point is that export the rest hander in mortar to the faas.

there have other methods to export the handler but I think is not gently, such as extra the mux in GRPCGatewayGeneratedHandlers.

**Comment by talgendler** [comment]
I think I understand your use case. You want to use unix sockets (UDS) to communicate, your Client and Server are on the same machine, right ?  

Mortar was built to support many different use case scenarios, but not all of them. Your case is a very specific one, but I still think it can be supported by providing a custom Listener. 
This [`providers.HTTPServerBuilderFxOption()`](https://github.com/airdb/xadmin-api/blob/dev/app/mortar/http.go#L19) line **Provides** a `partial.HTTPServerBuilder`. You can add an **Invoke** Fx Option to make changes to it.
Something like:
```go
providers.HTTPServerBuilderFxOption(),               // Web Server Builder
fx.Invoke(func(builder server.GRPCWebServiceBuilder) {
	builder.SetCustomListener(your listener)
}),
...
```
Order matters, the above "**Invoke**" addition must be called before calling [`BuildMortarWebServiceFxOption`](https://github.com/airdb/xadmin-api/blob/dev/cmd/api-server/main.go#L42) function which in turn calls the `Service` function in mortar where the actual call to `Build()` happens.

Even if my proposal is not going to work for you, your use case is still very unique and as such must be solved on your side. Adding an option to mutate the Builder doesn't feel right. Instead just right your own Builder. Mortar cares only about this [Interface](https://github.com/go-masonry/mortar/blob/master/interfaces/http/server/interfaces.go#L61).

Fork is always an option as well.
I'm sorry for the trouble I might cause asking you to provide a PR.

**Comment by xqbumu** [comment]
> I think I understand your use case. You want to use unix sockets (UDS) to communicate, your Client and Server are on the same machine, right ?
> 
> Mortar was built to support many different use case scenarios, but not all of them. Your case is a very specific one, but I still think it can be supported by providing a custom Listener. This [`providers.HTTPServerBuilderFxOption()`](https://github.com/airdb/xadmin-api/blob/dev/app/mortar/http.go#L19) line **Provides** a `partial.HTTPServerBuilder`. You can add an **Invoke** Fx Option to make changes to it. Something like:
> 
> ```go
> providers.HTTPServerBuilderFxOption(),               // Web Server Builder
> fx.Invoke(func(builder server.GRPCWebServiceBuilder) {
> 	builder.SetCustomListener(your listener)
> }),
> ...
> ```
> 
> Order matters, the above "**Invoke**" addition must be called before calling [`BuildMortarWebServiceFxOption`](https://github.com/airdb/xadmin-api/blob/dev/cmd/api-server/main.go#L42) function which in turn calls the `Service` function in mortar where the actual call to `Build()` happens.
> 
> Even if my proposal is not going to work for you, your use case is still very unique and as such must be solved on your side. Adding an option to mutate the Builder doesn't feel right. Instead just write your own Builder. Mortar cares only about this [Interface](https://github.com/go-masonry/mortar/blob/master/interfaces/http/server/interfaces.go#L61).
> 
> Fork is always an option as well. I'm sorry for the trouble I might cause asking you to provide a PR.

yes, we just want make our project has ability to deploy in serverless environment, and reduce the time cost in DevOps, so we need adapt some code for the serverless environment.
thanks for your replies. now I will close the pr, and feel free to reopen it, I am glad to see mortar has more gently and flexibly.

**Pull request closed without merging**